# Fiddling around with making Bungie API calls to get json data 

## [Error Codes](https://bungie-net.github.io/multi/schema_Exceptions-PlatformErrorCodes.html#schema_Exceptions-PlatformErrorCodes)
## [Manifest](https://bungie-net.github.io/multi/operation_get_Destiny2-GetDestinyManifest.html#operation_get_Destiny2-GetDestinyManifest)


In [179]:
import numpy as np
import pandas as pd
import requests
import os, sys
import re
import json
from bs4 import BeautifulSoup

module_path = os.path.abspath(os.path.join('..'))
if module_path not in sys.path:
    sys.path.append(module_path)
from seekerAlexaSkill.bungie import BungieData

### Trying to scrape meta weapons from Destiny Tracker

In [128]:
page = requests.get('https://destinytracker.com/') 
soup = BeautifulSoup(page.content, 'html.parser')
print(soup.prettify())

<!DOCTYPE html>
<html>
 <head>
  <script type="text/javascript">
   window.NREUM||(NREUM={});NREUM.info = {"beacon":"bam.nr-data.net","errorBeacon":"bam.nr-data.net","licenseKey":"5dfabc9f6c","applicationID":"100948194","transactionName":"ZlBSMhcEWRdWUxFQC18afTAmSn8LWlUmVgpFR18KCQBFS35eAVwc","queueTime":0,"applicationTime":6,"agent":"","atts":""}
  </script>
  <script type="text/javascript">
   window.NREUM||(NREUM={}),__nr_require=function(e,t,n){function r(n){if(!t[n]){var o=t[n]={exports:{}};e[n][0].call(o.exports,function(t){var o=e[n][1][t];return r(o||t)},o,o.exports)}return t[n].exports}if("function"==typeof __nr_require)return __nr_require;for(var o=0;o<n.length;o++)r(n[o]);return r}({1:[function(e,t,n){function r(){}function o(e,t,n){return function(){return i(e,[f.now()].concat(u(arguments)),t?null:this,n),t?void 0:this}}var i=e("handle"),a=e(2),u=e(3),c=e("ee").get("tracer"),f=e("loader"),s=NREUM;"undefined"==typeof window.newrelic&&(newrelic=s);var p=["setPageViewName","set

**Unfortunately, we can't get the meta list from Destiny Tracker since they're calling some asynchr method to input the info**

In [129]:
# Get PvP Meta weapons list
destinytracker.get_PvPMeta()

[<span class="name">{{ weapon.name }}</span>]


### Using the Bungie API for Destiny 2

In [182]:
cait = BungieData(api_key=os.environ["BUNGIE_API_KEY"], gamertag="BalancedSeeker6") # Never put your keys in code... export 'em!


In [183]:
# Get my own Bungie MembershipID
Me = cait.get_playerByTagName()
Me

[{'iconPath': '/img/theme/destiny/icons/icon_psn.png',
  'membershipType': 2,
  'membershipId': '4611686018453385415',
  'displayName': 'BalancedSeeker6'}]

In [184]:
# Get Destiny MembershipID
my_destiny_id = cait.get_DestinyUserId()
my_destiny_id

4611686018453385415

In [185]:
# Get my BUNGIE membershipID from using my DESTINY membershipID found in the above call
# This is the number found on the profile page of a user on Bungie.net
my_bungie_id = cait.get_BungieUserId()
my_bungie_id

11927477

### To use GetProfile, you'll need to lookup the user's Destiny membershipID by searching by gamertag. Found that in the get_DestinyUserId call : 4611686018453385415

In [186]:
My_Profile = cait.get_DestinyUserProfile(components=[100,200])
My_Profile

{'profile': {'data': {'userInfo': {'membershipType': 2,
    'membershipId': '4611686018453385415',
    'displayName': 'BalancedSeeker6'},
   'dateLastPlayed': '2018-06-15T07:07:38Z',
   'versionsOwned': 3,
   'characterIds': ['2305843009260518141',
    '2305843009289675361',
    '2305843009303434294']},
  'privacy': 1},
 'characters': {'data': {'2305843009260518141': {'membershipId': '4611686018453385415',
    'membershipType': 2,
    'characterId': '2305843009260518141',
    'dateLastPlayed': '2018-06-15T07:07:38Z',
    'minutesPlayedThisSession': '160',
    'minutesPlayedTotal': '12534',
    'light': 342,
    'stats': {'1935470627': 342,
     '2996146975': 3,
     '392767087': 9,
     '1943323491': 2},
    'raceHash': 2803282938,
    'genderHash': 2204441813,
    'classHash': 3655393761,
    'raceType': 1,
    'classType': 0,
    'genderType': 1,
    'emblemPath': '/common/destiny2_content/icons/60530e4d21f5a32bebdca82cb178c4bf.jpg',
    'emblemBackgroundPath': '/common/destiny2_cont

In [187]:
# Get a random single game's stats
game_stats = cait.get_postGameStats(100)
game_stats

{'period': '2017-09-01T17:50:42Z',
 'activityDetails': {'referenceId': 1658347443,
  'directorActivityHash': 1658347443,
  'instanceId': '100',
  'mode': 2,
  'modes': [7, 2],
  'isPrivate': False},
 'entries': [{'standing': 0,
   'score': {'basic': {'value': 0.0, 'displayValue': '0'}},
   'player': {'destinyUserInfo': {'iconPath': '/common/destiny2_content/icons/911791e90f955fc637398ea88aba74b7.jpg',
     'membershipType': 2,
     'membershipId': '4611686018468797828',
     'displayName': 'Mr_Angry'},
    'characterClass': 'Warlock',
    'classHash': 2271682572,
    'raceHash': 3887404748,
    'genderHash': 3111576190,
    'characterLevel': 1,
    'lightLevel': 100,
    'emblemHash': 1907674137},
   'characterId': '2305843009260533952',
   'values': {'assists': {'basic': {'value': 2.0, 'displayValue': '2'}},
    'completed': {'basic': {'value': 1.0, 'displayValue': 'Yes'}},
    'deaths': {'basic': {'value': 2.0, 'displayValue': '2'}},
    'kills': {'basic': {'value': 109.0, 'displayVa

In [188]:
manifest = cait.get_Manifest()
manifest

{'version': '64883.18.05.29.1341-7',
 'mobileAssetContentPath': '/common/destiny2_content/sqlite/asset/asset_sql_content_7012fd334ede7786ad089d9ceb9b20a5.content',
 'mobileGearAssetDataBases': [{'version': 0,
   'path': '/common/destiny2_content/sqlite/asset/asset_sql_content_7012fd334ede7786ad089d9ceb9b20a5.content'},
  {'version': 1,
   'path': '/common/destiny2_content/sqlite/asset/asset_sql_content_deb06e3cdc64eda41d665d489628c1b0.content'},
  {'version': 2,
   'path': '/common/destiny2_content/sqlite/asset/asset_sql_content_deb06e3cdc64eda41d665d489628c1b0.content'}],
 'mobileWorldContentPaths': {'en': '/common/destiny2_content/sqlite/en/world_sql_content_d63fda52ec4a1d85750480840e27534a.content',
  'fr': '/common/destiny2_content/sqlite/fr/world_sql_content_64bdcbcdc9907f64cb7ef7e296e415cf.content',
  'es': '/common/destiny2_content/sqlite/es/world_sql_content_8835e3094121015a7b2ca4cb8cda33aa.content',
  'de': '/common/destiny2_content/sqlite/de/world_sql_content_7dea966d758b3f8d

## Get a random statistic from Player Stats

In [189]:
# All time stats for PvE for all characters
my_stats['mergedAllCharacters']['results']['allPvE']['allTime'].keys()

dict_keys(['activitiesCleared', 'activitiesEntered', 'assists', 'totalDeathDistance', 'averageDeathDistance', 'totalKillDistance', 'kills', 'averageKillDistance', 'secondsPlayed', 'deaths', 'averageLifespan', 'bestSingleGameKills', 'bestSingleGameScore', 'opponentsDefeated', 'efficiency', 'killsDeathsRatio', 'killsDeathsAssists', 'objectivesCompleted', 'precisionKills', 'resurrectionsPerformed', 'resurrectionsReceived', 'score', 'heroicPublicEventsCompleted', 'adventuresCompleted', 'suicides', 'weaponKillsAutoRifle', 'weaponKillsFusionRifle', 'weaponKillsHandCannon', 'weaponKillsTraceRifle', 'weaponKillsPulseRifle', 'weaponKillsRocketLauncher', 'weaponKillsScoutRifle', 'weaponKillsShotgun', 'weaponKillsSniper', 'weaponKillsSubmachinegun', 'weaponKillsRelic', 'weaponKillsSideArm', 'weaponKillsSword', 'weaponKillsAbility', 'weaponKillsGrenade', 'weaponKillsGrenadeLauncher', 'weaponKillsSuper', 'weaponKillsMelee', 'weaponBestType', 'allParticipantsCount', 'allParticipantsScore', 'allParti

In [190]:
# All time stats for PvP for all characters
my_stats['mergedAllCharacters']['results']['allPvP']['allTime'].keys()

dict_keys(['activitiesEntered', 'activitiesWon', 'assists', 'totalDeathDistance', 'averageDeathDistance', 'totalKillDistance', 'kills', 'averageKillDistance', 'secondsPlayed', 'deaths', 'averageLifespan', 'score', 'averageScorePerKill', 'averageScorePerLife', 'bestSingleGameKills', 'bestSingleGameScore', 'opponentsDefeated', 'efficiency', 'killsDeathsRatio', 'killsDeathsAssists', 'objectivesCompleted', 'precisionKills', 'resurrectionsPerformed', 'resurrectionsReceived', 'suicides', 'weaponKillsAutoRifle', 'weaponKillsFusionRifle', 'weaponKillsHandCannon', 'weaponKillsTraceRifle', 'weaponKillsPulseRifle', 'weaponKillsRocketLauncher', 'weaponKillsScoutRifle', 'weaponKillsShotgun', 'weaponKillsSniper', 'weaponKillsSubmachinegun', 'weaponKillsRelic', 'weaponKillsSideArm', 'weaponKillsSword', 'weaponKillsAbility', 'weaponKillsGrenade', 'weaponKillsGrenadeLauncher', 'weaponKillsSuper', 'weaponKillsMelee', 'weaponBestType', 'winLossRatio', 'allParticipantsCount', 'allParticipantsScore', 'allP

In [191]:
play_type = np.random.choice(['allPvP', 'allPvE'])
play_stats = my_stats['mergedAllCharacters']['results'][play_type]['allTime']
print("Play Type: {}\n{}".format(play_type, play_stats))

Play Type: allPvE
{'activitiesCleared': {'statId': 'activitiesCleared', 'basic': {'value': 128.0, 'displayValue': '128'}}, 'activitiesEntered': {'statId': 'activitiesEntered', 'basic': {'value': 323.0, 'displayValue': '323'}}, 'assists': {'statId': 'assists', 'basic': {'value': 3039.0, 'displayValue': '3039'}, 'pga': {'value': 9.408668730650154, 'displayValue': '9.4'}}, 'totalDeathDistance': {'statId': 'totalDeathDistance', 'basic': {'value': 0.0, 'displayValue': '0.0'}}, 'averageDeathDistance': {'statId': 'averageDeathDistance', 'basic': {'value': 0.0, 'displayValue': '0.0'}}, 'totalKillDistance': {'statId': 'totalKillDistance', 'basic': {'value': 333020.0, 'displayValue': '333020.0'}}, 'kills': {'statId': 'kills', 'basic': {'value': 21228.0, 'displayValue': '21228'}, 'pga': {'value': 65.72136222910217, 'displayValue': '65.7'}}, 'averageKillDistance': {'statId': 'averageKillDistance', 'basic': {'value': 15.687770868664028, 'displayValue': '15.7'}}, 'secondsPlayed': {'statId': 'seconds

In [192]:
rand_stat_key = np.random.choice(list(play_stats.keys()))
rand_stat_key

'averageKillDistance'

In [193]:
' '.join(re.findall('[A-Z]?[a-z]+|[A-Z]+(?=[A-Z]|$)', rand_stat_key))

'average Kill Distance'

In [194]:
random_stat = play_stats[rand_stat_key]
random_stat

{'statId': 'averageKillDistance',
 'basic': {'value': 15.687770868664028, 'displayValue': '15.7'}}

In [195]:
random_stat['basic']['displayValue']

'15.7'

### Stat Definitions

In [196]:
stat_definitions = cait.get_StatDefinitions()
stat_definitions

{'activitiesCleared': {'statId': 'activitiesCleared',
  'group': 1,
  'periodTypes': [2, 1],
  'modes': [7, 6, 4, 2, 3, 16, 17, 46, 47, 18, 58],
  'category': 0,
  'statName': 'Activities Cleared',
  'statDescription': 'Successful completions of this mission type.',
  'unitType': 1,
  'unitLabel': '',
  'weight': 1},
 'activitiesEntered': {'statId': 'activitiesEntered',
  'group': 1,
  'periodTypes': [2, 1],
  'modes': [0,
   7,
   6,
   5,
   4,
   2,
   3,
   16,
   17,
   46,
   47,
   18,
   19,
   10,
   12,
   37,
   38,
   31,
   45,
   44,
   43,
   39,
   41,
   42,
   40,
   49,
   50,
   15,
   25,
   48,
   32,
   51,
   52,
   54,
   53,
   55,
   56,
   57,
   58,
   59],
  'category': 8,
  'statName': 'Activities Entered',
  'statDescription': 'Battles fought - some lost, some won.',
  'unitType': 1,
  'unitLabel': '',
  'weight': 1},
 'activitiesWon': {'statId': 'activitiesWon',
  'group': 1,
  'periodTypes': [2, 1],
  'modes': [5,
   19,
   10,
   12,
   37,
   38,
   

### Get Post Game Carnage Reports
>"If you want it specifically for you, you can use the *GetActivityHistory* endpoint to get all of your PGCR IDs (grab the **referenceId** property on the **DestinyHistoricalStatsActivity** objects being returned), and then use that to make calls to the *GetPostGameCarnageReport* endpoint!" - vthornheart-bng https://github.com/Bungie-net/api/issues/452


> "In the **PGCR** endpoint, there is a **referenceId**, which is the **activityHash** in **AggregateActivityStats** 
To get a player's match history, you need to use the GetActivityHistory endpoint  
GetActivityHistory = List of their recent matches  
GetPostGameCarnageReport = A specific match  
GetDestinyAggregateActivityStats = Aggregation of those matches per activityHash  - https://discordapp.com/channels/296008008956248066/296008136785920001/403851740686057472"

In [197]:
# Need a list of character IDs
My_Profile['profile']['data']['characterIds']

['2305843009260518141', '2305843009289675361', '2305843009303434294']

In [198]:
from collections import defaultdict

character_activities = {}

for character in My_Profile['profile']['data']['characterIds']:
    activity_ids = [act['activityDetails']['referenceId'] for act in cait.get_ActivityHistory(character, query="")['activities']]
    character_activities[character] = activity_ids

In [199]:
character_activities

{'2305843009260518141': [1435054848,
  3897312654,
  3233852802,
  806094750,
  2473919228,
  4012915511,
  2666761222,
  2473919228,
  2262757213,
  750001803,
  3164915257,
  2473919228,
  1003889713,
  148937731,
  990984849,
  1489679220,
  3089205900,
  2693136603,
  2693136603,
  451430877,
  958578340,
  2724706103,
  649648599,
  649648599,
  3676029623],
 '2305843009289675361': [1512980468,
  1057017675,
  399506119,
  750001803,
  2666761222,
  778271008,
  3143798436,
  778271008,
  806094750,
  399506119,
  1711620427,
  1711620427,
  3849796864,
  2666761222,
  1711620427,
  806094750,
  778271008,
  399506119,
  3164915257,
  3143798436,
  3143798436,
  2666761222,
  777592567,
  1583254851,
  806094750],
 '2305843009303434294': [4159221189,
  3143798436,
  4166562681,
  3966792859,
  564863404,
  2702476101,
  553537971,
  955852466,
  955852466,
  1243390694,
  3143798436,
  938512773,
  3143798436,
  3143798436,
  2926767881,
  3143798436,
  3143798436,
  2266954523,
 